File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/graphanalysis/HotPathAnalyzer.kt
/**
 * KSA-163: Hot Path Analyzer — Finds most-called functions.
 */
package com.codeintel.analyzers.graphanalysis

import com.codeintel.analyzers.graphanalysis.utils.GraphLoader

class HotPathAnalyzer(private val graphLoader: GraphLoader) {

    fun analyze(module: String? = null, limit: Int = 20, minCallers: Int = 2): List<HotPath> {
        val reverseGraph = graphLoader.loadReverseCallGraph(module)
        val results = mutableListOf<HotPath>()

        for ((symbolId, callers) in reverseGraph) {
            if (callers.size < minCallers) continue
            val transitiveCallers = computeTransitiveCallers(symbolId, reverseGraph)
            val info = graphLoader.getSymbolInfo(symbolId) ?: continue
            results.add(HotPath(
                symbolId = symbolId, symbolName = info.name,
                filePath = info.filePath, directCallers = callers.size,
                transitiveCallers = transitiveCallers, kind = info.kind,
            ))
        }
        return results.sortedByDescending { it.transitiveCallers }.take(limit)
    }

    private fun computeTransitiveCallers(symbolId: Int, reverseGraph: AdjacencyList): Int {
        val visited = mutableSetOf(symbolId)
        val queue = ArrayDeque<Int>()
        queue.addLast(symbolId)
        while (queue.isNotEmpty()) {
            val current = queue.removeFirst()
            for (caller in reverseGraph[current] ?: emptyList()) {
                if (caller !in visited) {
                    visited.add(caller); queue.addLast(caller)
                }
            }
        }
        return visited.size - 1
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/graph_analysis/hot_path_analyzer.py
"""KSA-163: Hot Path Analyzer."""
from __future__ import annotations
from collections import deque
from . import HotPath, AdjacencyList
from .utils import GraphLoader


class HotPathAnalyzer:
    def __init__(self, graph_loader: GraphLoader) -> None:
        self._gl = graph_loader

    def analyze(self, module: str | None = None, limit: int = 20, min_callers: int = 2) -> list[HotPath]:
        rg = self._gl.load_reverse_call_graph(module)
        results = []
        for sid, callers in rg.items():
            if len(callers) < min_callers: continue
            tc = self._transitive_callers(sid, rg)
            info = self._gl.get_symbol_info(sid)
            if not info: continue
            results.append(HotPath(sid, info.name, info.file_path, len(callers), tc, info.kind))
        return sorted(results, key=lambda h: -h.transitive_callers)[:limit]

    def _transitive_callers(self, sid: int, rg: AdjacencyList) -> int:
        visited = {sid}
        queue: deque[int] = deque([sid])
        while queue:
            cur = queue.popleft()
            for c in rg.get(cur, []):
                if c not in visited:
                    visited.add(c); queue.append(c)
        return len(visited) - 1
